### PR TITLE
[feature] new implementation of Quick sort

### PR DESCRIPTION
**Pull request by kvedala**
#### Description of Change

This is a revival of the PR #876 that seems to have been abandoned by its original author.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/958"></a>

### CODE REVIEW

File: sorting/quick_sort_3.cpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+ * @author immortal-j
+ * @author [Krishna Vedala](https://github/kvedala)
+ */
+#include <algorithm>
+#include <cassert>
+#include <ctime>
+#include <iostream>
+#include <vector>
+
+namespace {
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:18]: Undocumented namespace.

File: sorting/quick_sort_3.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            std::swap((*arr)[high], (*arr)[low]);
+        }
+        *i = low;
+        *j = high;
+        return;
+    }
+
+    int32_t mid = low;
+    T pivot = (*arr)[high];
+    while (mid <= high) {
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:70]: Undocumented.

File: sorting/quick_sort_3.cpp [hunk 3]
```diff
@@ -0,0 +1,188 @@
+/**
+ * @file
+ * @brief Implementation Details
+ * @details Quick sort 3 works on Dutch National Flag Algorithm
+ * The major difference between simple quicksort and quick sort 3 comes in the
+ * function partition3 In quick_sort_partition3 we divide the vector/array into
+ * 3 parts. quick sort 3 works faster in some cases as compared to simple
+ * quicksort.
+ * @author immortal-j
```
Comment (Panquesito7) [inline sorting/quick_sort_3.cpp:9]:
```suggestion
 * @author [immortal-j](https://github.com/immortal-j)
```

File: sorting/quick_sort_3.cpp (hunk 1 above)
Comment (kvedala) [reply sorting/quick_sort_3.cpp:18]: it is meant to be. please learn about un-named namespaces

File: sorting/quick_sort_3.cpp (hunk 2 above)
Comment (kvedala) [reply sorting/quick_sort_3.cpp:70]: ? what is to be documented here?

File: sorting/quick_sort_3.cpp (hunk 3 above)
Comment (kvedala) [reply sorting/quick_sort_3.cpp:9]: this was authors' contribution. will leave it as is.

Comment (Panquesito7) [reply sorting/quick_sort_3.cpp:9]: Still not resolved...

**Pull request merged**